README.md
To run this code, the following packages need to be installed:


changelog.md
#### 11/20/2022 (Peter)
    1. Added functionality to handle .csv files.
    2. Added functionality to dynamically export results.
        - The names are the same as the imported names.
    3. Ran results on the "combined_hits.csv" file with ~581 extracted peptides.
    4. More general cleanup/commenting to code.
    
#### 11/21/2022 (Nivedha)
    1. Changed the file name to inference_with_hits
    2. Added Positive Hits ipynb file
    3. It takes the combined csv file gives out peptides sorted based on confidence score

#### 11/26/2022 (Peter)
    1. Recreated the regression models with scikit-learn version 1.1.2 properly installed.
    2. Renamed them to the proper format (.pkl).

#### 11/28/2022 (Peter)
    1. Split the "inference.py" script into three scripts:
        a. "inference.py" holds the inference functions.
        b. "pipeline.py" holds the model pipeline
        c. "main.py" calls our entire pipeline/inference
    2. More general clean-up.

#### 12/1/2022 (Peter)
    1. Cleaned up the file imports in pipeline.py
    2. pipeline.py changes migrated to inference.ipynb

#### 12/17/2022 (Peter)
    1. Placed the dependencies and generated files in their own folders to clean up the directory before exporting to github.
    2. Created a github repo for this project

#### 12/26/2022 (Peter)
    1. Updated "pipeline.py"
        a. Changed the features included and models used to be the RFE models (see below).
        b. Added functionality for PCA in the model pipeline.
        c. Added global variable "VARIANCE" that is used to determine how many principal components to select.
            - This current version uses 90% variance.
    2. Updated "Regression Dependencies" wit the following (did not remove old models and dependencies):
        a. Added "Lasso Regression trained model medium bucket (rfe).pkl"
        b. Added "Lasso Regression trained model small bucket (rfe).pkl"
        c. Added "rfe_selected_features.json"
        d. Added "SVC with Linear Kernel 10.00 rfe-pca.pkl"
        e. Added "SVC with Linear Kernel trained model (rfe).pkl"

#### 12/29/2022 (Peter)
    1. Updated "Regression Dependencies"
        a. Added "SVR with Linear Kernel trained model medium bucket (rfe).pkl"
        b. Added "SVR with Linear Kernel trained model small bucket (rfe).pkl"
        c. Added "SVR with RBF Kernel trained model medium bucket (rfe).pkl"
        d. Added "SVR with RBF Kernel trained model small bucket (rfe).pkl"
    2. Updated "pipeline.py"
        a. Added notes to the code.

#### 1/2/2023 (Peter)
    1. Fixed bug generating a Lasso model instead of SVR with RBF Kernel for the trained medium/small (rfe) bucket models.


#### 3/17/2023 (Peter)
    1. Created new branch 'reg_only_update' that has updates for the regression only pipeline.  Merge the branches once regression only section is done.

#### 3/31/2023 (Peter)
    1. Changed the variance on the regression only model to 85%.

#### 5/1/2023 (Peter)
    1. Changed the models for the Regression only dependencies with fixed data.

#### 5/3/2023 (Peter)
    1. Retrained the model to exclude any peptides with KI values > 4000

#### 5/5/2023 (Peter)
    1. Reran inference function on new file.

#### 5/7/2023 (Nivedha)
    1. Changed the threshold level from 0.25 to 0.4 (line 65 on pipeline.py)
    2. Removed ensemble functionality

pipeline.py
## Package Imports
# Unpack Files
import json
import pickle

# Data Analysis
import pandas as pd
import numpy as np 

# Change this global variable depending on what variance we choose for PCA.  Set this to 'False' if we don't end up using PCA.
REGRESSION_ONLY_VARIANCE = False
BUCKET_VARIANCE = False
KI_RANGE = (-11.330603908176274, 17.19207365866807)
SOURCE_INTERVAL = (-5,5)

## Pipeline for multi-stage model:
def model_pipeline(allFeaturesData):
    """
    Function for the multistage pipeline.  The following stages are applied:
        - Classification of the peptides.  Positive classification means efficacy in antithrombin response.
        - Classification of the positive peptides into buckets.  0 = small, 1 = medium, 2 = large.
        - Regression on peptides that are set to buckets 0 and 1, to predict "KI (nM)" values for the peptides.

    Parameters
    ----------
    allFeaturesData: Pandas DataFrame containing the peptides with extracted features.

    ensemble: Boolean Flag to determine use of an ennsemble method for the initial peptide classification.

    Returns:
    --------
    result: Pandas DataFrame containing the results with columns ['Name','Seq','Predicted','KI (nM) Predicted']
    """
    ### MODEL IMPORTS
    # ------------------------------------
    ## Classification Model Imports:
    # Scaler
    with open('Classification Dependencies/Scaler Classification.pkl', 'rb') as fh:
        scaler_for_classification = pickle.load(fh)

    # Trained Classification Model
    with open('Classification Dependencies/SVM Linear Classification trained.pkl', 'rb') as fh:
        classification_model_1 = pickle.load(fh)

    # Selected features Classification
    with open('Classification Dependencies/Features for Classification Model.json') as fh:
        classification_features = json.loads(fh.read())

    ### CLASSIFICATION
    # ----------------------------------

    # get only necessary features
    clf_data = allFeaturesData[classification_features]

    # Apply preprocessing function
    clf_data = pd.DataFrame(scaler_for_classification.transform(clf_data),
                                                        columns = clf_data.columns)

    # Model 1 - Applying threshold on decision function
    decisionFuctionModel1 = classification_model_1.decision_function(clf_data)
    threshold = 0.4

    y_predict = []
    for j in decisionFuctionModel1:
        if j > threshold:
            y_predict.append(1)
        else:
            y_predict.append(0)

    predictions =  pd.DataFrame(y_predict, columns=["Predicted"])
    predictions.replace({1:'Positive',0:'Negative'}, inplace=True)
    allFeaturesData = pd.concat([allFeaturesData, predictions], axis=1)
    
    ### REGRESSION
    # ----------------------------------

    # get positively predicted peptides
    reg_data = pd.DataFrame(allFeaturesData[allFeaturesData['Predicted']=='Positive'], columns = allFeaturesData.columns)
    
    # Exception if no peptides are predicted positive.
    if len(reg_data) == 0:
         # save result in a new dataframe
        result = allFeaturesData[['Name','Seq','Predicted']]
        return result

    y_predict_bucketed = bucket_regression_inference_pipeline(reg_data)
    y_predict_unbucketed = regression_only_inference_pipeline(reg_data)

    reg_data['KI (nM) Predicted (bucketized)'] = y_predict_bucketed
    reg_data['KI (nM) Predicted (non-bucketized)'] = y_predict_unbucketed


    allFeaturesData = pd.merge(allFeaturesData,reg_data[['Seq', 'KI (nM) Predicted (bucketized)',
                                                         'KI (nM) Predicted (non-bucketized)']],                                                     
                               on='Seq', how='left')

    # save result in a new dataframe
    result = allFeaturesData[['Name','Seq','Predicted', 'KI (nM) Predicted (bucketized)',
                              'KI (nM) Predicted (non-bucketized)']]

    return result

def bucket_regression_inference_pipeline(reg_data=pd.DataFrame()):
    # Scaler
    with open('Regression Dependencies/Scaler Regression.pkl', 'rb') as fh:
        regression_scaler = pickle.load(fh)

    # Classification with ki
    with open('Regression Dependencies/SVC RBF bucket classification trained.pkl', 'rb') as fh:
        bucket_classification_model = pickle.load(fh)

    # Regression for Medium Bucket
    with open('Regression Dependencies/SVR RBF medium bucket Regression trained.pkl', 'rb') as fh:
        regression_model_medium_bucket = pickle.load(fh)

    # Regression for Small Bucket
    with open('Regression Dependencies/SVR RBF small bucket Regression trained.pkl', 'rb') as fh:
        regression_model_small_bucket = pickle.load(fh)

    # Selected features Regression
    with open('Regression Dependencies/Features for Regression Model.json') as fh:      
        # Changed this back to sequential forward selection.
        regression_features = json.loads(fh.read())

    # Apply preprocessing function and select only necessary features
    reg_data = pd.DataFrame(regression_scaler.transform(reg_data.iloc[:,2:-1]), 
                                                columns = reg_data.columns[2:-1])[regression_features]
    
    if BUCKET_VARIANCE != False:
        with open('Regression Dependencies/SVC with Linear Kernel 10.00 rfe-pca.pkl', 'rb') as fh:
            pca = pickle.load(fh)
        reg_data = pd.DataFrame(pca.transform(reg_data))

        # Dimensionality Reduction based on accepted variance.
        ratios = np.array(pca.explained_variance_ratio_)
        ratios = ratios[ratios.cumsum() <= (BUCKET_VARIANCE/100)]

        # Readjust the dimensions of 'reg_data' based on the variance we want.
        length = len(ratios)
        if length > 0:
            reg_data = reg_data[reg_data.columns[0:length]]

    # Predict the buckets
    buckets_pred = bucket_classification_model.predict(reg_data)
    reg_data['Bucket'] = buckets_pred

    # Make predictions for all of the buckets.  The large bucket is predict as 0.  Only make predictions if the
    #   arrays aren't empty.
    if reg_data[buckets_pred==0].size != 0:
        sml_pred = regression_model_small_bucket.predict(reg_data[buckets_pred==0].iloc[:,:-1])
        sml_pred = np.exp(np.interp(sml_pred, SOURCE_INTERVAL, KI_RANGE))
    if reg_data[buckets_pred==1].size != 0:
        med_pred = regression_model_medium_bucket.predict(reg_data[buckets_pred==1].iloc[:,:-1])
        med_pred = np.exp(np.interp(med_pred, SOURCE_INTERVAL, KI_RANGE))
    lrg_pred = np.zeros(np.count_nonzero(reg_data[buckets_pred==2]))

    # Put back the predictions in the original order.
    y_predict_regression = np.array([])
    for i in buckets_pred:
        if i == 0:
            y_predict_regression = np.append(y_predict_regression, sml_pred[0])
            sml_pred = np.delete(sml_pred, 0)
        elif i == 1:
            y_predict_regression = np.append(y_predict_regression, med_pred[0])
            med_pred = np.delete(med_pred, 0)
        elif i == 2:
            y_predict_regression = np.append(y_predict_regression, lrg_pred[0])
            lrg_pred = np.delete(lrg_pred, 0)

    return y_predict_regression

def regression_only_inference_pipeline(reg_data=pd.DataFrame()):
    """
    This function handles the regression only portion of the inference pipeline if we never had a 
        classification stage in the pipeline.
        
    Parameters
    ----------
    reg_data: Pandas DataFrame containing the table of data for the various peptides in the positive
        dataset.

    Regurns
    -------
    result: A single column containing the predicted KI values 
    """

    ## MODEL IMPORTS
    # MinMaxScaler Transformation
    with open('Regression Only Dependencies/regression only scaler.pkl', 'rb') as fh:
        regression_scaler = pickle.load(fh)

    # sequential Forward Selection Features Selected
    with open('Regression Only Dependencies/regression_only_selected_features.json', 'rb') as fh:
        regression_features = json.loads(fh.read())

    # Regression model
    with open('Regression Only Dependencies/regression only SVR with RBF Kernel trained model.pkl', 'rb') as fh:
        regression_model = pickle.load(fh)

    # Scaler transformation and Feature Selection
    reg_data = pd.DataFrame(regression_scaler.transform(reg_data.iloc[:,2:-1]),
                                    columns = reg_data.columns[2:-1])[regression_features]
    # Apply PCA if needed.
    if REGRESSION_ONLY_VARIANCE != False:
        with open('Regression Only Dependencies/PCA for SVR with RBF Kernel.pkl', 'rb') as fh:
            pca = pickle.load(fh)
        reg_data = pd.DataFrame(pca.transform(reg_data))

        # Dimensionality Reduction based on accepted variance.
        ratios = np.array(pca.explained_variance_ratio_)
        ratios = ratios[ratios.cumsum() < (REGRESSION_ONLY_VARIANCE/100)]

        # Readjust the dimensions of 'reg_data' based on the bariance we get.
        length = len(ratios)
        if length > 0:
            reg_data = reg_data[reg_data.columns[0:length]]

    # Predictions
    y_pred = regression_model.predict(reg_data)
    y_pred = np.exp(np.interp(y_pred, SOURCE_INTERVAL, KI_RANGE))

    return y_pred

github_update.sh
# Folders
git add __pycache__
git add .ipynb_checkpoints
git add 'Classification Dependencies'
git add 'Regression Dependencies'
git add 'Regression Only Dependencies'
git add 'Results'

# Markdown Files
git add changelog.md
git add README.md

# Python Scripts
git add main.py
git add inference.py
git add pipeline.py
git add inference_with_hits.py

# Jupyter Notebooks
git add Inference_with_hits.ipynb
git add Inference.ipynb
git add 'Positive Hits.ipynb'

# Bash Scripts
git add github_update.sh

# Fasta Files
git add longhits.fasta
git add smallhits.fasta

# .csv files
git add combined_hits.csv
git add single_results.csv
git add sorted_peptides.csv
git add medoid_peptides.csv

# .fai files
git add longhits.fasta.fai
git add smallhits.fasta.fai

# .txt files
git add requirements.txt

git commit -m '5/7/2023 - updated classification models and removed ensemble model'

inference_with_hits.py
## Dependency Imports:
# Compute protein descriptors
from propy import PyPro
from propy import AAComposition
from propy import CTD

# Build Sequence Object
from Bio.SeqUtils.ProtParam import ProteinAnalysis

# Read Fasta File
from pyfaidx import Fasta

# Grouping iterable
from itertools import chain

# Return file path
import glob

# Unpack Files
import json
import pickle

# Dataframes
import pandas as pd
import numpy as np

## Classification Model Imports:
# Scaler
file = open('Scaler Classification.pkl', 'rb')
scaler_for_classification = pickle.load(file)
file.close()

# First Model
file = open('SVM Linear Classification trained.pkl', 'rb')
classification_model_1 = pickle.load(file)
file.close()

# Second Model
file = open('SVM RBF Classification trained.pkl', 'rb')
classification_model_2 = pickle.load(file)
file.close()

# Selected features Classification
file = open("Features for Classification Model.json")
classification_features = json.loads(file.read())
file.close()

## Regression Model Imports:
# Scaler
file = open('Scaler Regression.pickle', 'rb')
scaler_for_regression = pickle.load(file)
file.close()

# Classification with ki
file = open('SVC RBF bucket Classification trained.pickle', 'rb')
classification_model_for_buckets = pickle.load(file)
file.close()

# Regression for Medium Bucket
file = open('SVR RBF medium bucket Regression trained.pickle', 'rb')
regression_model_medium_bucket = pickle.load(file)
file.close()

# Regression for Small Bucket
file = open('SVR RBF small bucket Regression trained.pickle', 'rb')
regression_model_small_bucket = pickle.load(file)
file.close()

# Selected features Regression
file = open("Features for Regression Model.json")
regression_features = json.loads(file.read())
file.close()

## Pipeline for multi-stage model:
def model_pipeline(allFeaturesData, ensemble=bool):
    """
    Function for the multistage pipeline.  The following stages are applied:
        - Classification of the peptides.  Positive classification means efficacy in antithrombin response.
        - Classification of the positive peptides into buckets.  0 = small, 1 = medium, 2 = large.
        - Regression on peptides that are set to buckets 0 and 1, to predict "KI (nM)" values for the peptides.

    Parameters
    ----------
    allFeaturesData: Pandas DataFrame containing the peptides with extracted features.

    ensemble: Boolean Flag to determine use of an ennsemble method for the initial peptide classification.

    Returns:
    --------
    result: Pandas DataFrame containing the results with columns ['Name','Seq','Predicted','KI (nM) Predicted']
    """
    # CLASSIFICATION
    # ----------------------------------

    # get only necessary features
    clf_data = allFeaturesData[classification_features]

    # Apply preprocessing function
    clf_data = pd.DataFrame(scaler_for_classification.transform(clf_data),
                                                        columns = clf_data.columns)

    # Model 1 - Applying threshold on decision function
    decisionFuctionModel1 = classification_model_1.decision_function(clf_data)
    threshold = 0.25
    y_predict_model_1 = []
    for j in decisionFuctionModel1:
        if j > threshold:
            y_predict_model_1.append(1)
        else:
            y_predict_model_1.append(0)

    # Model 2
    y_predict_model_2 = classification_model_2.predict(clf_data)

    # Ensemble Model Prediction.  If the flag is set to 'True', we combine results of SVC with RBF + Linear Kernels.
    if ensemble==True:
        y_predict_ensemble = []
        for i in range(len(y_predict_model_1)):
            if (y_predict_model_1[i] == 1) & (y_predict_model_2[i] == 1):
                y_predict_ensemble.append('Positive')
            else:
                y_predict_ensemble.append('Negative')
        allFeaturesData = pd.concat([allFeaturesData, pd.DataFrame(y_predict_ensemble, columns=["Predicted"])], axis=1)

    elif ensemble==False:
        predictions_model_2 =  pd.DataFrame(y_predict_model_2, columns=["Predicted"])
        predictions_model_2.replace({1:'Positive',0:'Negative'}, inplace=True)
        allFeaturesData = pd.concat([allFeaturesData, predictions_model_2], axis=1)
    
    # REGRESSION
    # ----------------------------------

    # get positively predicted peptides
    reg_data = pd.DataFrame(allFeaturesData[allFeaturesData['Predicted']=='Positive'], columns = allFeaturesData.columns)
    
    # Exception for negative peptides
    if len(reg_data) == 0:
         # save result in a new dataframe
        result = allFeaturesData[['Name','Seq','Predicted']]
        return result
        
    # Apply preprocessing function and select only necessary features
    reg_data_reduced = pd.DataFrame(scaler_for_regression.transform(reg_data.iloc[:,2:-1]), 
                                                        columns = reg_data.columns[2:-1])[regression_features]

    # Predict the buckets.
    buckets_pred = classification_model_for_buckets.predict(reg_data_reduced)
    reg_data_reduced['Bucket'] = buckets_pred

    # Fixed Ki range and Source Interval
    ki_range = (-11.330603908176274, 17.19207365866807)
    source_interval = (-5,5)

    # Make predictions for all of the buckets. The large bucket is predict as 0. Only make predictions if the arrays aren't empty.
    if reg_data_reduced[buckets_pred==0].size != 0:
        sml_pred = regression_model_small_bucket.predict(reg_data_reduced[buckets_pred==0].iloc[:,:-1])
        sml_pred = np.exp(np.interp(sml_pred, source_interval, ki_range))
    if reg_data_reduced[buckets_pred==1].size != 0:
        med_pred = regression_model_medium_bucket.predict(reg_data_reduced[buckets_pred==1].iloc[:,:-1])
        med_pred = np.exp(np.interp(med_pred, source_interval, ki_range))
    lrg_pred = np.zeros(np.count_nonzero(reg_data_reduced[buckets_pred==2]))

    # Put back the predictions in the original order.
    y_predict_regression = np.array([])
    for i in buckets_pred:
        if i == 0:
            y_predict_regression = np.append(y_predict_regression, sml_pred[0])
            sml_pred = np.delete(sml_pred, 0)
        elif i == 1:
            y_predict_regression = np.append(y_predict_regression, med_pred[0])
            med_pred = np.delete(med_pred, 0)
        elif i == 2:
            y_predict_regression = np.append(y_predict_regression, lrg_pred[0])
            lrg_pred = np.delete(lrg_pred, 0)

    reg_data['KI (nM) Predicted'] = y_predict_regression

    allFeaturesData = pd.merge(allFeaturesData,reg_data[['Seq','KI (nM) Predicted']],on='Seq', how='left')

    # save result in a new dataframe
    result = allFeaturesData[['Name','Seq','Predicted','KI (nM) Predicted']]

    return result

    ## Inference Function for Fasta File:
def inferenceFasta(fastafile, ensemble=True):
    
    """ The inference function gets the protein sequence, trained model, preprocessing function and selected
    features as input. 

    The function read the sequence as string and extract the peptide features using appropriate packages into 
    the dataframe.

    The necessary features are selected from the extracted features which then undergoes preprocessing function, the
    target value is predicted using trained function and give out the results. """
    
    new_peptides = []
    for file in glob.glob(fastafile):
        new_peptides.append(file)
        
    for f in new_peptides:
        fa = Fasta(f)
        # empty list to save the features
        allFeaturesData = []
        for seq in fa:
            # Make sure the sequence is a string
            s = str(seq)
            
            # replace the unappropriate peptide sequence to A
            s = s.replace('X','A')
            s = s.replace('x','A')
            s = s.replace('U','A')
            s = s.replace('Z','A')
            s = s.replace('B','A')

            # Calculating primary features
            analysed_seq = ProteinAnalysis(s)
            wt = analysed_seq.molecular_weight()
            arm = analysed_seq.aromaticity()
            instab = analysed_seq.instability_index()
            flex = analysed_seq.flexibility()
            pI = analysed_seq.isoelectric_point()

            # create a list for the primary features
            pFeatures = [seq.name, s, len(seq), wt, arm, instab, pI]

            # Get Amino Acid Composition (AAC), Composition Transition Distribution (CTD) and Dipeptide Composition (DPC)
            resultAAC = AAComposition.CalculateAAComposition(s)
            resultCTD = CTD.CalculateCTD(s)
            resultDPC = AAComposition.CalculateDipeptideComposition(s)

            # Collect all the features into lists
            aacFeatures = [j for i,j in resultAAC.items()]
            ctdFeatures = [l for k,l in resultCTD.items()]
            dpcFeatures = [n for m,n in resultDPC.items()]
            allFeaturesData.append(pFeatures + aacFeatures + ctdFeatures + dpcFeatures)
        
        # Collect feature names
        pFeaturesName = ['Name','Seq' ,'SeqLength','Weight','Aromaticity','Instability','IsoelectricPoint']
        aacFeaturesData = [i for i,j in resultAAC.items()]
        ctdFeaturesData = [k for k,l in resultCTD.items()]
        dpcFeaturesData = [m for m,n in resultDPC.items()]
        
        featuresName  = []
        featuresName.append(pFeaturesName+aacFeaturesData+ctdFeaturesData+dpcFeaturesData)
        featuresFlattenList = list(chain.from_iterable(featuresName))
        
        # create dataframe using all extracted features and the names
        allFeaturesData = pd.DataFrame(allFeaturesData, columns = featuresFlattenList)
        
        result = model_pipeline(allFeaturesData, ensemble)

        return result

## Inference Function for Single Sequences:
def inferenceSingleSeqence(seq, ensemble=True):
    
    """ The inference function gets the protein sequence, trained model, preprocessing function and selected
    features as input. 
    
    The function read the sequence as string and extract the peptide features using appropriate packages into 
    the dataframe.
    
    The necessary features are selected from the extracted features which then undergoes preprocessing function, the
    target value is predicted using trained function and give out the results. """
    
    # empty list to save the features
    allFeaturesData = []
    
    # Make sure the sequence is a string
    s = str(seq)
    
    # replace the unappropriate peptide sequence to A
    s = s.replace('X','A')
    s = s.replace('x','A')
    s = s.replace('U','A')
    s = s.replace('Z','A')
    s = s.replace('B','A')
    
    # Calculating primary features
    analysed_seq = ProteinAnalysis(s)
    wt = analysed_seq.molecular_weight()
    arm = analysed_seq.aromaticity()
    instab = analysed_seq.instability_index()
    flex = analysed_seq.flexibility()
    pI = analysed_seq.isoelectric_point()
    
    # create a list for the primary features
    pFeatures = [seq, s, len(seq), wt, arm, instab, pI]
     
    # Get Amino Acid Composition (AAC), Composition Transition Distribution (CTD) and Dipeptide Composition (DPC)
    resultAAC = AAComposition.CalculateAAComposition(s)
    resultCTD = CTD.CalculateCTD(s)
    resultDPC = AAComposition.CalculateDipeptideComposition(s)
    
    # Collect all the features into lists
    aacFeatures = [j for i,j in resultAAC.items()]
    ctdFeatures = [l for k,l in resultCTD.items()]
    dpcFeatures = [n for m,n in resultDPC.items()]
    allFeaturesData.append(pFeatures + aacFeatures + ctdFeatures + dpcFeatures)
    
    # Collect feature names
    name1 = ['Name','Seq' ,'SeqLength','Weight','Aromaticity','Instability','IsoelectricPoint']
    name2 = [i for i,j in resultAAC.items()]
    name3 = [k for k,l in resultCTD.items()]
    name4 = [m for m,n in resultDPC.items()]
    name  = []
    name.append(name1+name2+name3+name4)
    flatten_list = list(chain.from_iterable(name))
    
    # create dataframe using all extracted features and the names
    allFeaturesData = pd.DataFrame(allFeaturesData, columns = flatten_list)

    result = model_pipeline(allFeaturesData, ensemble)

    return result

## Inference Function for .csv Files:
def inference_csv(csv, ensemble=True):
    sequences = pd.read_csv('combined_hits.csv')
    sequences = sequences.replace(r"^ +| +$", r"", regex=True)
    sequences = sequences['Seq']

    sequence_data = []
    for seq in sequences:
        # Make sure the sequence is a string
        s = str(seq)
        
        # replace the unappropriate peptide sequence to A
        s = s.replace('X','A')
        s = s.replace('x','A')
        s = s.replace('U','A')
        s = s.replace('Z','A')
        s = s.replace('B','A')

        # Calculating primary features
        analysed_seq = ProteinAnalysis(s)
        wt = analysed_seq.molecular_weight()
        arm = analysed_seq.aromaticity()
        instab = analysed_seq.instability_index()
        flex = analysed_seq.flexibility()
        pI = analysed_seq.isoelectric_point()

        # create a list for the primary features
        pFeatures = [seq, s, len(seq), wt, arm, instab, pI]

        # Get Amino Acid Composition (AAC), Composition Transition Distribution (CTD) and Dipeptide Composition (DPC)
        resultAAC = AAComposition.CalculateAAComposition(s)
        resultCTD = CTD.CalculateCTD(s)
        resultDPC = AAComposition.CalculateDipeptideComposition(s)

        # Collect all the features into lists
        aacFeatures = [j for i,j in resultAAC.items()]
        ctdFeatures = [l for k,l in resultCTD.items()]
        dpcFeatures = [n for m,n in resultDPC.items()]
        sequence_data.append(pFeatures + aacFeatures + ctdFeatures + dpcFeatures)

    # Collect feature names
    pFeaturesName = ['Name','Seq' ,'SeqLength','Weight','Aromaticity','Instability','IsoelectricPoint']
    aacFeaturesData = [i for i,j in resultAAC.items()]
    ctdFeaturesData = [k for k,l in resultCTD.items()]
    dpcFeaturesData = [m for m,n in resultDPC.items()]

    featuresName  = []
    featuresName.append(pFeaturesName+aacFeaturesData+ctdFeaturesData+dpcFeaturesData)
    featuresFlattenList = list(chain.from_iterable(featuresName))

    # create dataframe using all extracted features and the names
    sequence_data = pd.DataFrame(sequence_data, columns = featuresFlattenList)

    result = model_pipeline(sequence_data, ensemble)

    return result

## Implementation Function:
def inference(file=str, ensemble=True):
    if file.endswith('.fasta') == True:
        result = inferenceFasta(file, ensemble)
        name = file.split(sep='.')[0]
        result.to_csv('%s_fasta_results.csv' %(name))
    elif file.endswith('.csv') == True:
        result = inference_csv(file, ensemble)
        name = file.split(sep='.')[0]
        result.to_csv('%s_csv_results.csv' %(name))
    else:
        result = inferenceSingleSeqence(file, ensemble)
        result.to_csv('single_results.csv')
    return result

inference('combined_hits.csv', ensemble=True)

inference('APEADQTTPEEKPAEPEPVA', ensemble=True)

inference('APEADQTTPEEKPAEPEPVA', ensemble=False)

inference('QSPLPERQE', ensemble=True)

inference('QSPLPERQE', ensemble=False)

inference('HTLGYINDNEEGPR', ensemble=True)

inference('HTLGYINDNEEGPR', ensemble=False)

inference('smallhits.fasta', ensemble=True)

inference('smallhits.fasta', ensemble=False)

pd.set_option('display.max_rows',None)
inference('longhits.fasta', ensemble=True)

pd.set_option('display.max_rows',None)
inference('longhits.fasta', ensemble=False)